### ISSUE TITLE
pkg_resources is deprecated

### ISSUE THREAD

**Issue by luhn** [Pylons/pyramid#3731]
As of [setuptools v67.5.0](https://setuptools.pypa.io/en/latest/history.html#v67-5-0), a deprecation warning is raised when importing pkg_resources:

> DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html

Apparently this has been replaced by [importlib.resources](https://docs.python.org/3/library/importlib.resources.html) and [importlib.metadata](https://docs.python.org/3/library/importlib.metadata.html), which is available since Python 3.7, as well as [packaging](https://packaging.pypa.io/en/stable/).

This may be fairly tricky to remove, as `src/pyramid/config/assets.py` digs pretty deep into the pkg_resources API, namely it registers `PackageOverrides` with `pkg_resources.register_loader_type`.  However I can't figure out *why* this call is necessary, I can't find where this loader would be used—Maybe something to do with `AssetResolver`?

The other issue is that pkg_resources would unzip the whole package into a tmp directory when calling `resource_filename`, whereas importlib just extracts that single file.  `pyramid.path.package_path` depends on this behavior.  Possible solutions:

- Deprecate `pyramid.path.package_path`, and by extension `pramid.path.caller_path`, `pyramid.asset.asset_spec_from_abspath`, and `pyramid.resource.resource_path_from_abspath`.
- Change contract so `pyramid.path.package_path` and it's dependents don't work with zip files (or returning zip "path")
- Write an implementation that extracts package to tmp directory, mimicking pkg_resources behaviior.

**Comment by mmerickel** [comment]
The issue is around `config.override_asset` which overrides how `pkg_resources` works on a threadlocal basis such that any `pkg_resources` (and `AssetResolver`) invocations find the appropriate packages. When I have looked, `importlib.resources` does not have hooks to implement this functionality and is going to be a sticking point in what Pyramid can do about this issue without requiring setuptools.

I'm not really aware of places that importlib.metadata would be helpful in pyramid but those have been very easy to replace in any package where I've done it to drop setuptools for that purpose.

**Comment by luhn** [comment]
It may be possible to override importlib machinery?  A custom `Loader` can provide a `get_resource_reader` method, which then can return a `ResourceReader` object.   https://docs.python.org/3.8/library/importlib.html#setting-up-an-importer

Do we *want* to monkeypatch the import machinery, though?  It seems to go against Pyramid's MO of everything contained within a registry and discouraging threadlocals.  Now's the time to break from that, as all projects will have to rewrite their `pkg_resources` usage anyways.

As a counterpoint, however, maintainers will by default follow the [migration guide](https://importlib-resources.readthedocs.io/en/latest/migration.html) and expect it to just work.

**Comment by mmerickel** [comment]
I’m gonna be honest I don’t really know what to do about this. It’s been in the back of my mind for a long time and just haven’t seen a path forward. 

- Pyramid could offer its own apis that support the override and we remove the support. Release a patch version that deprecated things so folks can get the warnings. 

- Pyramid could completely remove asset override support. 

- Pyramid could try to patch the loaders and support folks overriding importlib.resources. Keeping the current pkg_resources support but making it a soft dependency instead of hard. 

Am I missing an option?

**Comment by luhn** [comment]
As a compromise between one and two, we could keep override support for static views but drop support for global overrides.  Static overrides I think are necessary, as you often have static assets stored outside of your Python package, but I'm not convinced how useful it is outside of that.. Perhaps even actively harmful:  Can you imagine trying to debug why your app is loading a resource that is *not* the one in the package, all because of a single `override_asset` buried somewhere in the codebase? 

To ease the transition, we could have a `pkg_resources` shim that uses `importlib.resource` by default, but if `override_asset` is called it the shim switches to `pkg_resources`.  That way any apps relying on overrides will still work for now, while other apps can avoid the deprecation warnings and be unaffected when pkg_resources is eventually removed.  A new `override_static_asset` method (or perhaps an `override` kwarg for `add_static_view`?) would provide static-only overrides with no monkeypatching of the import machinery.

**Comment by mmerickel** [comment]
The override feature is very useful because it allows some libraries to easily provide tweaks. For example pypi itself used it (not sure if still) to open source most of their ui while keeping parts of it closed because they were owned by the graphic designer. The override allowed a simple theme concept for them. 

Also as I said before it’s expected to work for all of the templating libraries for a similar reason.

I guess my point is that it’s definitely not there just for static views.

**Comment by mmerickel** [comment]
I'm genuinely at a loss how to deal with this issue. I want to do something to make pkg_resources optional but it's so integral to a some core apis that currently expect a threadlocal place to do a lookup.

I think we'd have to make the asset resolution apis work "correctly" without pkg_resources, and then offer an extra api to opt-in to binding the data to pkg_resources if someone wants it.

Working correctly is the challenge there - basically it means we'd do some form of asset lookup / registry ourselves that can obey `config.override_asset` to preserve that feature via `get_current_registry()` instead of going straight to pkg_resources.

**Comment by mmerickel** [comment]
We'd then update other libraries like pyramid_jinja2 and pyramid_mako to use whatever API instead. Right now pyramid_jinja2 uses `abspath_from_asset_spec` which isn't even a public pyramid API right now.

Of course one option is to just make things like `abspath_from_asset_spec` work as expected without using `pkg_resources` by starting its lookup at some `get_current_registry().queryUtility(IAssetRegistry)` that controls the lookup via `importlib.resources`, `pkg_resources`, etc.

**Comment by mmerickel** [comment]
I have removed the easy spots in #3748 and #3749. Down to 4 locations that are impacted:

- `pyramid.assets.abspath_from_asset_spec`
- `config.override_asset` installs the `pkg_resources.DefaultProvider` hooks
- `pyramid.path.PkgResourcesAssetDescriptor`
- `pyramid.static` uses direct apis that should probably all be replaced with `pyramid.path.AssetResolver`.

Ultimately the focus is:

- How to make `pyramid.path.AssetResolver` work with a new `IAssetDescriptor`, and "do the right thing" by respecting `config.override_asset()`.
- How to still allow something like `config.install_pkg_resources_asset_provider()`
- Fix everything within pyramid to use `pyramid.path.AssetResolver`.

**Comment by mmerickel** [comment]
Related, I opened this conversation with importlib_resources after diving into this a bit.

https://github.com/python/importlib_resources/issues/295

**Comment by robvdl** [comment]
Just an observation, but the project seems to be in a transitional phase and stuff is spread out across pyproject.toml, setup.py and setup.cfg which is quite confusing. But Pyramid being an older project that totally makes sense.

My guess is that eventually the setup.py and setup.cfg will be dropped.

**Comment by mmerickel** [comment]
It's true we will at some point port most of our project metadata into `pyproject.toml` but this ticket isn't about how we're using setuptools to build Pyramid. It's specifically about using pkg_resources within the actual pyramid package.

**Comment by stevepiercy** [comment]
FYI, Plone started getting hit with this in Python 3.13, as well as in documentation builds where `conf.py` still uses it. For reference and perhaps save y'all some effort to hunt down replacements.

- https://github.com/plone/plone.api/pull/566/files
- https://github.com/plone/Products.CMFPlone/issues/4126

**Comment by john-sivak** [comment]
@mmerickel Hello, have you been able to make more progress on addressing pkg_resources deprecation?
I see you had some successful dialog with the importlib_resources folks here: https://github.com/python/importlib_resources/issues/295

Thanks

**Comment by mmerickel** [comment]
@ChrisBarker-NOAA progress has stalled because of my lack of time/motivation to focus on the work. I would love to try to get it working but I haven't felt the need from the community to really find the work valuable.

The complexities are around:
1. The fact that the existing APIs are designed around setuptools which exposes "global" resources. The new importlib requires context managers instead. This causes problems within our AssetResolver and static_view implementations that need workarounds to handle properly.
2. Asset overriding in importlib - they merged the PR so it should be possible but requires an entirely new threadlocal impl in Pyramid.
3. Syncing changes to things like pyramid_jinja2 which are also using pkg_resources.

**Comment by fschulze** [comment]
I made a PR for pyramid_chameleon: https://github.com/Pylons/pyramid_chameleon/pull/38
For the devpi-web project the only issue now is ``static_view``. I will look into that next, but I think that is "just" a chore.
I have no clue yet what the asset overriding stuff is about. Maybe we should make a point release which deprecates it and see who complains? They will get another wakeup call in December when setuptools drops pkg_resources.

**Comment by ChrisBarker-NOAA** [comment]
@nandoflorestan: yes, that's what this thread is about -- we all want to remove the dependency, but someone(s) need to find teh time to do it.

IIUC, it's pretty straightforward, except for the use in  config.override_asset -- and some progress has been made.

Meanwhile, it IS the right thing for that deprecation warning to show up!

(which is why showing deprecation warnings if off by default in the python interpreter)

**Comment by adamk550** [comment]
Like others I am waiting for an update on this issue. We have a bunch of modules we use, some of which depend on setuptools. Right now using a setuptools version < 82.0.0 is fine as nothing _requires_ a later version but when that does happen we won't be able to upgrade things

My main concern is security vulnerabilities: if something is found in one of our dependencies we need to upgrade and if that means moving to something that needs setuptools 82+ then are stuck. No one wants to be in that situation.

**Comment by luhn** [comment]
Hi all, @mmerickel and I had a long discussion to hash out the details.  We now have a roadmap for removing setuptools as a dependency.  Any apps/add-ons that still depend on setuptools/pkg_resources will be expected to specify it as a dependency—If pkg_resources is present, Pyramid will automatically register overrides with it to preserve legacy behavior.

Phase 1:  Introduce new APIs.

- Add `Configuration.resolve_asset` and `Request.resolve_asset` APIs.
- Add documentation to "What's New" on how to migrate from `pkg_resources` to new asset APIs.
- Provide an alternative to `pkg_resources.resource_filename` (importlib.resource only provides filenames within context manager) — WIP at https://github.com/Pylons/pyramid/pull/3791/
- Document `IAssetDescriptor.abspath` as discouraged (but not deprecated)
- Implement `FSAssetDescriptor.absspec`, returning same value as `abspath`.
- Add `IAssetDescriptor.origspec`; modify `IAssetDescriptor.absspec` to return post-override spec.  **TBD**
- Update `pyramid.response.FileResponse` to handle asset specs (via `abspath`), not just file paths.

Phase 2:  Migrate away from `pkg_resources` internally

- Move asset mechanisms to `importlib.resources`   — WIP at https://github.com/Pylons/pyramid/pull/3791/
	- Implement `ImportlibResourcesAssetDescriptor` (alias to `PkgResourcesAssetDescriptor` for bw compat) utilizing `IPackageOverrides`
	- Remove `pkg_resources` from `IPackageOverrides` and `PackageAssetSource`, register `pkg_resources` loader separately based on presence of `pkg_resources`.
- Migrate `pyramid.static.static_view` to new asset APIs — WIP at https://github.com/Pylons/pyramid/pull/3750
- Migrate `pyramid.config.views.StaticURLInfo._bust_asset_path` to new asset APIs (utilizing `origspec` and `absspec`)
- Migrate `pyramid.asset.abspath_from_asset_spec` to new asset APIs — WIP at https://github.com/Pylons/pyramid/pull/3750
- Migrate `pyramid.path.package_path` to new asset APIs.

Phase 3:  Remove `setuptools` as a dependency.

Phase 4 (optional):  Deprecations

- `Configurator.absolute_asset_spec` — Can use `config.resolve_asset(path).absspec()` instead.  Exception:  Doesn't have passthrough behavior for non-strings.
- `pyramid.path.caller_path` (no dependents) — Can use `AssetResolver().resolve(path).abspath()` instead.
- `pyramid.path.resolve_asset_spec (depends `static_view`, `_bust_asset_path`, `absolute_asset_spec`, should be refactored away in phase 2) — Can use `AssetResolver().resolve(path).absspec()` instead.  Returns string instead of two-tuple.
- `pyramid.asset.asset_spec_from_abspath` (no dependents except already deprecated `pyramid.resource.resource_spec_from_abspath`) — Doesn't have an alternative in new APIs.
- `pyramid.path.package_path` (no dependents except previous `caller_path` and `asset_spec_from_abspath`, depends on `pkg_resources`) — Can use `AssetResolver(package).resolve('').abspath()` instead.

**Comment by jvanasco** [comment]
> Any apps/add-ons that still depend on setuptools/pkg_resources will be expected to specify it as a dependency

Is there a known list of any official/popular add-ons that will require this yet, or will this be "see what breaks!" situation?

**Comment by mmerickel** [comment]
If pyramid addons or app code want to protect themselves then they have two choices:
- Depend on setuptools < 82 directly.
- Use `pyramid.path.AssetResolver` instead of `pkg_resources`. It's a drop-in replacement that does and will continue to work.

Obviously the latter is better - pyramid_jinja2 for example is doing this. It'd be great to get a list of popular addons that are directly using pkg_resources so they can be improved.

Also of course if you aren’t relying on asset overrides at all you can just use importlib.resources directly today - but addons shouldn’t do that. Only app code is safe enough to make that assumption.

**Comment by luhn** [comment]
I’ve done some searching across the pylons GitHub and pyramid_chameleon was the only offender I found. Probably wouldn’t be too much to do a manual sweep of https://trypyramid.com/extending-pyramid.html before release.

**Comment by jvanasco** [comment]
`pyramid_debugtoolbar` is clean.

`pyramid_mailer` also docs:

* https://github.com/search?q=repo%3APylons%2Fpyramid_mailer%20pkg_resources&type=code

`pyramid_mako` uses it for docs:

* https://github.com/search?q=repo%3APylons%2Fpyramid_mako%20pkg_resources&type=code

`pyramid_tm`, docs and tests

* https://github.com/search?q=repo%3APylons%2Fpyramid_tm%20pkg_resources&type=code

On the above, it's just used to get the version of the installed package.  

That can be replaced with importlib.metadata, since the min pyramid version is 3.10 (and that's 3.8+)

I don't think those will break installs, but maybe we could do a single migration ticket here and crossrefernce/checklist it against those projects (and pyramid_jinja2)

**Comment by mmerickel** [comment]
> I don't think those will break installs

Yeah none of those are a runtime dependency issue, they just need to be fixed in order to be able to publish new versions of the libraries. I don't want that type of stuff tracked here as it's unrelated.

**Comment by Daverball** [comment]
Considering this issue now blocks upgrading `setuptools` to a version, that does not have a security vulnerability, you probably need to revise your 4 phase rollout plans, considering we have not even made it to phase 2 yet and we won't get rid of the dependency until phase 3.

Would it perhaps be possible, as a first step, that can be done right now, to vendor `pkg_resources` into an internal namespace, so the runtime dependency on `setuptools` can be removed? Or is there some `pkg_resources` internal registry that would need to be synced with the `setuptools` version of the package, if an older version of `setuptools` is installed? But perhaps that can easily be solved by the vendored module automatically replacing its internal registry with the `setuptools` version, if it detects that there is `pkg_resources` module in the global namespace?

That seems like a good interim solution, that leaves the ball in people's courts and leaves people with enough time to migrate to the new internal APIs.

**Comment by luhn** [comment]
All the phases are done over at https://github.com/Pylons/pyramid/compare/main...luhn:pyramid:asset-project2, it's just a matter of chunking it into smaller PRs and getting those merged.  I have #3815 and #3816 open currently.

**Comment by fmigneault** [comment]
Are there any updates / soon intention to integrate the open PRs?
Given [CVE-2026-59890](https://www.cve.org/CVERecord?id=CVE-2026-59890) affecting `setuptools<83`, it is starting to be inconvenient to pin `setuptools<82` just for `pyramid` to work.
